wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/model/impl/devfile/DevfileImpl.java

```diff
@@ -93,34 +95,41 @@ public DevfileImpl(
     if (attributes != null) {
       this.attributes = new HashMap<>(attributes);
     }
+
+    if (metadata != null) {
+      this.metadata = new MetadataImpl(metadata);
+    }
+  }
+
+  @PrePersist
+  @PreUpdate
+  public void validate() {
+    if (metadata == null || metadata.getName() == null || metadata.getName().isEmpty()) {
+      throw new IllegalStateException("The devfile must have metadata with a non-empty name.");
```

Comment:
Can you elaborate more why did you decide to put such validate on `@PrePersist` and `@PreUpdate`?

I'm asking because we have another place where we do validation, and should we duplicate it here?
If yes, then should not we duplicate another validation, like required `apiVersion`, or required name of a project?

Reply:
Well, if there wasn't a step like this, we would get an SQL error. The reason I put it in here was merely to provide a nicer error message than a constraint error from the DB.

If the devfile object comes through the intended codepaths (e.g. from the REST API down to the business logic), it would have been validated against schema, which should disallow it from reaching this point anyway.

Reply:
> Well, if there wasn't a step like this, we would get an SQL error.

The same if apiVersion is missing, no? My point - do for all or for nothing.

Reply:
I removed the check so that we only rely on the schema validation.